afs_mission_goal/pipeline/usoc/creating_datasets/02_derive_variables.py

```diff
@@ -0,0 +1,322 @@
+    ]
+    informal_vars_13 = [
+        "cchours10",
+        "cchours11",
+        "cchours12",
+        "cchours13",
+        "cchours16",
+        "cchours17",
+    ]
+
+    ukhls_master = process_ukhls_data(
+        ukhls_master=ukhls_master,
+        formal_vars_15=formal_vars_15,
+        informal_vars_15=informal_vars_15,
+        formal_vars_612=formal_vars_612,
+        informal_vars_612=informal_vars_612,
+        formal_vars_13=formal_vars_13,
+        informal_vars_13=informal_vars_13,
+    )
+
+    # CREATE CATEGORICAL VARIABLE ##
+    # follow SEED definitions FORMAL CHILDCARE https://assets.publishing.service.gov.uk/media/5e4e5c10e90e074dcd5bd213/SEED_AGE_5_REPORT_FEB.pdf
+    # high use - using over 20 hours per week
+    #  medium use - using 10-20 hours per week
+    # low use - using 0-10 hours per week
+    # not using - not using childcare
+
+    # Apply the categorization function
+    ukhls_master["formal_cchrs_cat"] = ukhls_master["formal_hours"].apply(
+        categorize_formal_hours
+    )
+
+    # Map the numerical categories to descriptive labels
+    formal_cchrs_labels = {
+        0: "no formal childcare use",
+        1: "1-10 hrs formal childcare use",
+        2: "11-20 hrs formal childcare use",
+        3: "20+ hours formal childcare use",
+    }
+
+    ukhls_master["formal_cchrs_cat"] = ukhls_master["formal_cchrs_cat"].map(
+        formal_cchrs_labels
+    )
+
+    # Convert 'wave' variable from letter to number (assuming 'wave' is a letter)
+    ukhls_master["wave"] = ukhls_master["wave"].apply(lambda x: ord(x) - ord("a") + 1)
+
+    # Create 'interview date' and 'dob date' variables
+    ukhls_master["int_date"] = create_date_columns(
+        ukhls_master, ["intdaty_dv", "intdatm_dv"]
+    )
+    ukhls_master["dob_date"] = create_date_columns(ukhls_master, ["doby_dv", "dobm_dv"])
+
+    # Add the new variables to the master_vars DataFrame
+    newvars = pd.DataFrame(
+        {
+            "origin": [np.nan] * len(ukhls_master.columns),
+            "variable": ukhls_master.columns.astype(str),
+        }
+    )
+
+    # Filter newvars to exclude variables in oldvars
+    newvars = newvars[~newvars["variable"].isin(oldvars)]
+
+    # Define the variable groups
+    incomevars = ["equivinc"] + [var for var in newvars["variable"] if "real" in var]
+    benefitvars = [var for var in newvars["variable"] if "ben" in var]
+    childcarevars = [var for var in newvars["variable"] if "formal" in var]
+
+    # Update the origin column based on conditions
+    newvars["origin"] = np.select(
+        [
+            newvars["variable"].isin(incomevars),
+            newvars["variable"].isin(benefitvars),
+            newvars["variable"].isin(childcarevars),
+            newvars["variable"] == "int_date",
+            newvars["variable"] == "dob_date",
+        ],
+        ["hhresp", "indresp", "child", "hhresp", "xwave"],
+        default=np.nan,
```

Comment:
This doesn't run for me, because the types don't work. It runs if you replace this with `pd.NA`.
```suggestion
        default=pd.NA,
```